File: 06_semestre/ttps/problems/45_light_more_light.cpp
#include <iostream>
#include <vector>
#include <cmath>

const int MAX_LIMIT = 65536;
bool is_prime[MAX_LIMIT + 1];
std::vector<int> primes;

void sieve_of_eratosthenes() {
    std::fill(is_prime, is_prime + MAX_LIMIT + 1, true);
    is_prime[0] = is_prime[1] = false;
    for (int i = 2; i <= std::sqrt(MAX_LIMIT); i++) {
        if (is_prime[i]) {
            for (int j = i * i; j <= MAX_LIMIT; j += i) {
                is_prime[j] = false;
            }
        }
    }
    for (int i = 2; i <= MAX_LIMIT; i++) {
        if (is_prime[i]) {
            primes.push_back(i);
        }
    }
}

long long divisors_p(long long n) {
    long long total = 1;
    for (int p : primes) {
        if (static_cast<long long>(p) * p > n) {
            break;
        }
        long long curr = 0;
        while (n % p == 0) {
            curr++;
            n /= p;
        }
        total *= curr + 1;
    }
    return total * (n > 1 ? 2 : 1);
}

int main() {
    sieve_of_eratosthenes();
    std::string line;
    while (std::getline(std::cin, line)) {
        long long n = std::stoll(line);
        if (n < 2) {
            if (n == 0) {
                break;
            }
            std::cout << "yes\n";
            continue;
        }
        if ((divisors_p(n) % 2) == 0) {
            std::cout << "no\n";
        } else {
            std::cout << "yes\n";
        }
    }
    return 0;
}

File: 06_semestre/ttps/problems/45_light_more_light.py
"""
https://onlinejudge.org/index.php?option=com_onlinejudge&Itemid=8&page=show_problem&problem=1051

There is man named ”mabu” for switching on-off light in our University. He
switches on-off the lights in a corridor. Every bulb has its own toggle switch.
That is, if it is pressed then the bulb turns on.
Another press will turn it off. To save power consumption (or may be he is mad
or something else) he does a peculiar thing. If in a corridor there is n bulbs,
he walks along the corridor back and forth n times and in i-th walk he toggles
only the switches whose serial is divisable by i. He does not press any switch
when coming back to his initial position. A i-th walk is defined as going down
the corridor (while doing the peculiar thing) and coming back again. Now you
have to determine what is the final condition of the last bulb. Is it on or off?

Input
The input will be an integer indicating the n'th bulb in a corridor. Which is
less then or equals 2^32 -1. A zero indicates the end of input. You should not
process this input.

Output
Output 'yes' if the light is on otherwise 'no', in a single line.

Sample Input
3
6241
8191
0

Sample Output
no
yes
no
**ENDS IN NEW LINE**
"""

import math
import sys


MAX_POSSIBLE_PRIME = math.ceil(math.sqrt(2**32 - 1))


def sieve_of_eratosthenes(limit: int):
    is_prime = [True] * (limit + 1)
    is_prime[0] = False
    is_prime[1] = False
    for i in range(2, int(math.sqrt(limit)) + 1):
        if is_prime[i]:
            for i in range(i * i, limit + 1, i):
                is_prime[i] = False

    return [p for p in range(2, limit + 1) if is_prime[p]]


def divisors(n: int) -> int:
    total = 1

    for p in primes:
        if p * p > n:
            break

        curr = 0

        while n % p == 0:
            curr += 1
            n //= p

        total *= curr + 1

    return total * (2 if n > 1 else 1)


primes = sieve_of_eratosthenes(MAX_POSSIBLE_PRIME)

lines = sys.stdin.read().split()
r = []  # type: list[str] # type: ignore

for line in lines:
    n = int(line)
    if n < 2:
        if n == 0:
            break

        r.append("yes\n")
        continue

    if (divisors(n) % 2) == 0:
        r.append("no\n")
    else:
        r.append("yes\n")

sys.stdout.writelines(r)
